### PR TITLE
Topic/cstr varargs array part2

### PR DESCRIPTION
**Pull request by soc**

### CODE REVIEW

File: dora/src/bytecode/generator.rs [hunk 1]
```diff
@@ -1548,6 +1549,13 @@ impl<'a, 'ast> AstBytecodeGen<'a, 'ast> {
 
                 Intrinsic::GenericArrayNew => self.emit_intrinsic_new_array(expr, dest),
 
+                Intrinsic::GenericArrayWithValues => {
+                    let ty = self.ty(expr.id);
+                    let array_reg = self.emit_array_with_variadic_arguments(expr, &[ty], 0, dest);
+                    self.registers.free_if_temp(array_reg);
```
Comment (dinfuehr) [inline dora/src/bytecode/generator.rs:1555]: I don't think this line is correct.

File: dora/src/bytecode/generator.rs [hunk 2]
```diff
@@ -1580,24 +1588,11 @@ impl<'a, 'ast> AstBytecodeGen<'a, 'ast> {
...
-        let callee = self.vm.fcts.idx(callee_id);
-        let callee = callee.read();
-
-        // Determine types for arguments and return values
-        let (arg_types, _, return_type) = self.determine_callee_types(&call_type, &*callee);
-
-        assert!(return_type.is_unit());
-        assert_eq!(arg_types.len(), 2);
+        // We need array of elements
+        let element_ty = self.ty(expr.id);
```
Comment (dinfuehr) [inline dora/src/bytecode/generator.rs:1592]: Nice!

File: dora/src/bytecode/generator.rs (hunk 1 above)
Comment (soc) [reply dora/src/bytecode/generator.rs:1559]: So this line needs to happen after the caller got the register (and likely did something with it), right?

Comment (dinfuehr) [reply dora/src/bytecode/generator.rs:1507]: I think you can simply remove that line. array_reg can't be freed at that point, it should be freed by the caller later. Unfortunately there is no check in the bytecode generator yet that makes sure that instruction operands are still live and not freed.

Comment (dinfuehr) [reply dora/src/bytecode/generator.rs:1507]: Btw the caller should already free the register, so you shouldn't need to change any other code.

Comment (soc) [reply dora/src/bytecode/generator.rs:1507]: I think without that line I ran into the assert – I'll check again.

Comment (soc) [reply dora/src/bytecode/generator.rs:1507]: Weeeeird – it works! I would have sworn that didn't work earlier ...

Comment (dinfuehr) [reply dora/src/bytecode/generator.rs:1507]: That was without the dest argument, right?

**Comment by soc** [comment]
@dinfuehr This should be fine now, right?

Could you merge https://github.com/dinfuehr/dora/pull/227 before this one, so I can rebase it cleanly?

**Comment by soc** [comment]
Oops, yes, sorry. Will fix.

**Comment by soc** [comment]
@dinfuehr Fixed!

**Pull request merged**